### PR TITLE
Check for duplicate modules

### PR DESCRIPTION
**Pull request by ripleyaffect**
Throws an error for:
- Duplicate UUIDs or module codes
- Duplicate UUIDs for checkpoints within the same module

Result:
image

### CODE REVIEW

File: src/lib/expandModuleCheckpoints.js [hunk 1]
```diff
@@ -20,11 +20,13 @@ module.exports = async (module, libraryFiles) => {
...
 
-    // Ensure the checkpoint is valid
-    await validateCheckpoint(checkpoint);
+    checkpoints.push(checkpoint);
+  }
 
-    chekpoints.push(checkpoint);
+  // Ensure checkpoints are valid
+  for (const checkpoint of checkpoints) {
+    await validateCheckpoint(checkpoint, checkpoints);
```
Comment (learykara) [inline src/lib/expandModuleCheckpoints.js]: what's the purpose of the changes in this file? same end result, no?

File: src/lib/validateCheckpoint.js [hunk 2]
```diff
@@ -10,6 +14,17 @@ module.exports = async checkpoint => {
...
   }
 
+  // Check for a unique uuid
+  const checkpointsWithUuid = checkpoints.filter(
+    cp => cp.attributes.uuid === checkpoint.attributes.uuid
+  );
+  if (checkpointsWithUuid.length > 1) {
+    error =
+      `CHECKPOINT_INVALID: ${checkpointsWithUuid.length} ` +
+      `checkpoints found with uuid "${checkpoint.attributes.uuid}": ` +
```
Comment (learykara) [inline src/lib/validateCheckpoint.js]: 😱 you're not modifying the plurality of `checkpoints` based on `checkpointsWithUuid.length` ? who are you even?

File: src/lib/validateCheckpoint.js [hunk 3]
```diff
@@ -10,6 +14,17 @@ module.exports = async checkpoint => {
...
+      `${formatDuplicates(checkpointsWithUuid)} - Expected 1`;
```
Comment (learykara) [inline src/lib/validateCheckpoint.js]: expected 0, no?

File: src/lib/validateModule.js [hunk 4]
```diff
@@ -10,6 +16,27 @@ module.exports = async mod => {
...
+    error = `MODULE_INVALID: Module "${mod.src}" has no name`;
+  }
+
+  // Check for a unique code
+  const modsWithCode = mods.filter(m => m.code === mod.code);
+  if (modsWithCode.length > 1) {
+    error =
+      `MODULE_INVALID: ${modsWithCode.length} ` +
+      `modules found with code "${mod.code}": ` +
+      `${formatDuplicates(modsWithCode)} - Expected 1`;
```
Comment (learykara) [inline src/lib/validateModule.js]: ditto

File: src/lib/validateCheckpoint.js (hunk 2 above)
Comment (ripleyaffect) [reply src/lib/validateCheckpoint.js]: it has to be >1 to hit this

File: src/lib/validateCheckpoint.js (hunk 3 above)
Comment (ripleyaffect) [reply src/lib/validateCheckpoint.js]: expected ONLY one, because the current checkpoint included in the list

File: src/lib/expandModuleCheckpoints.js (hunk 1 above)
Comment (ripleyaffect) [reply src/lib/expandModuleCheckpoints.js]: all checkpoints are put into the list before validation, otherwise only the previous ones would be checked as duplicates...which I guess would also work technically?

Comment (ripleyaffect) [reply src/lib/expandModuleCheckpoints.js]: I need to move this to something that validates the entire library anyway
